### PR TITLE
Add Sensorless Homing current

### PR DESCRIPTION
**Pull request by danym21**
### Requirements
TMC drivers with stallguard enabled

### Description
I use TMC5160 stepper drivers, since I use 0.9deg/step instead 1.8deg/step motors stallguard is not triggered correctly with the the current that I use for printing.

as example for printing I use a current from 1100 mA and "STALL_SENSITIVITY 2" the machine is crashing, when I use "STALL_SENSITIVITY 1" Stallguard is trigger often in the middle of the way to the edge. When I use a current from 462 for homing with "STALL_SENSITIVITY 2" all is working fine but on printing it is losing steps.

### Benefits
in the homing process set a different current on the steppers so that the sensorless homing is 
more flexible.
also spreadCycle can be used as option.

### Related Issues

[FR] different driver current for sensorless_homing #15686

### CODE REVIEW

**Comment by thisiskeithb** [comment]
Have you tried homing faster? Sensorless homing with StallGuard tends to work better if things are moving quickly.

**Comment by danym21** [comment]
shure I have tried this, it crash on the machine much harder but will also not help when the head is very near the zero point, then there is not enough time to accelerate.

File: Marlin/src/gcode/calibrate/G28.cpp [hunk 1]
```diff
@@ -263,42 +256,40 @@ void GcodeSuite::G28(const bool always_home_all) {
...
-  #if ENABLED(USE_HOMING_CURRENT)
-    #ifdef X_CURRENT_HOME
-      TMC_SAVE_CURRENT_X = stepperX.getMilliamps();
-      SERIAL_ECHOLN(TMC_SAVE_CURRENT_X);
+  #define HAS_CURRENT_HOME(N) (defined(N##_CURRENT_HOME) && N##_CURRENT_HOME != N##_CURRENT)
+  #define HAS_HOMING_CURRENT (HAS_CURRENT_HOME(X) || HAS_CURRENT_HOME(X2) || HAS_CURRENT_HOME(Y) || HAS_CURRENT_HOME(Y2))
+
+  #if HAS_HOMING_CURRENT
+    auto debug_current = [](const char * const s, const int16_t a, const int16_t b){
+      DEBUG_ECHO(s); DEBUG_ECHOLNPGM(" current: ", a, " -> ", b);
```
Comment (danym21) [inline Marlin/src/gcode/calibrate/G28.cpp:265]: I think this is not working, maybe change DEBUG_ECHOLNPGM to DEBUG_ECHOLNPAIR

File: Marlin/src/gcode/calibrate/G28.cpp [hunk 2]
```diff
@@ -263,42 +256,40 @@ void GcodeSuite::G28(const bool always_home_all) {
...
   #endif
 
-  #if ENABLED(STEALTHCHOP_XY)
-    #if ENABLED(USE_SPREADCYCLE_FOR_HOMING)
-      if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPGM("change driver XY to spreadCycle");
-      process_subcommands_now_P(PSTR("M569 S0 X Y"));
-    #endif
+  #if BOTH(STEALTHCHOP_XY, HOME_USING_SPREADCYCLE)
+    if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPGM("Set XY to spreadCycle...");
+    process_subcommands_now_P(PSTR("M569S0XY"));
```
Comment (danym21) [inline Marlin/src/gcode/calibrate/G28.cpp:291]: spaces needed: process_subcommands_now_P(PSTR("M569 S0 X Y"));

File: Marlin/src/gcode/calibrate/G28.cpp [hunk 3]
```diff
@@ -520,32 +511,27 @@ void GcodeSuite::G28(const bool always_home_all) {
...
-  
-  #if ENABLED(STEALTHCHOP_XY)
-    #if ENABLED(USE_SPREADCYCLE_FOR_HOMING)
-      if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPGM("change driver XY back to StealthChop");
-      process_subcommands_now_P(PSTR("M569 S1 X Y"));
-    #endif
+
+  #if BOTH(STEALTHCHOP_XY, HOME_USING_SPREADCYCLE)
+    if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPGM("Set XY to StealthChop...");
+    process_subcommands_now_P(PSTR("M569S1XY"));
```
Comment (danym21) [inline Marlin/src/gcode/calibrate/G28.cpp:522]: spaces needed: process_subcommands_now_P(PSTR("M569 S0 X Y"));

**Comment by thinkyhead** [comment]
I'm working on cleaning up and standardizing all the endstops code, including sensorless with DIAG0 and SPI-based sensorless. I notice that homing current was added to configurations for Z2, etc., but it is not implemented in the code. And later Z homing current was changed to only apply to `DELTA`. So, these are some issues I hope to work out in PR #25758 before it gets merged. Please add your commentary to that PR if you are still interested in this topic.

**Pull request merged**